core/src/apps/cardano/address.py

```diff
@@ -1,137 +1,249 @@
-from trezor import log
-from trezor.crypto import base58, crc, hashlib
-
-from apps.cardano.protocol_magics import ProtocolMagics
-from apps.common import HARDENED, cbor
+from trezor import wire
+from trezor.crypto import base58, hashlib
+from trezor.messages import CardanoAddressParametersType, CardanoAddressType
+
+import apps.cardano.address_id as AddressId
+from apps.cardano import BYRON_PURPOSE, SHELLEY_PURPOSE
+from apps.cardano.bech32 import bech32_encode
+from apps.cardano.byron_address import derive_byron_address
+from apps.cardano.utils import variable_length_encode
+from apps.common import HARDENED
 from apps.common.seed import remove_ed25519_prefix
 
 if False:
-    from typing import Tuple
-    from trezor.crypto import bip32
+    from typing import List
+    from trezor.messages import CardanoBlockchainPointerType
     from apps.cardano import seed
 
 
-def _encode_address_raw(address_data_encoded):
-    return base58.encode(
-        cbor.encode(
-            [cbor.Tagged(24, address_data_encoded), crc.crc32(address_data_encoded)]
-        )
-    )
+BECH32_ADDRESS_PREFIX = "addr"
 
 
-def derive_address_and_node(
-    keychain: seed.Keychain, path: list, protocol_magic: int
-) -> Tuple[str, bip32.Node]:
-    node = keychain.derive(path)
+def validate_full_path(path: List[int]) -> bool:
+    """
+    Validates derivation path to fit {44', 1852'}/1815'/a'/{0,1,2}/i,
+    where `a` is an account number and i an address index.
+    The max value for `a` is 20, 1 000 000 for `i`.
+    """
+    if len(path) != 5:
+        return False
+    if path[0] != BYRON_PURPOSE and path[0] != SHELLEY_PURPOSE:
+        return False
+    if path[1] != 1815 | HARDENED:
+        return False
+    if path[2] < HARDENED or path[2] > 20 | HARDENED:
+        return False
+    if path[3] != 0 and path[3] != 1 and path[3] != 2:
+        return False
+    if path[4] > 1000000:
+        return False
+    return True
+
+
+def matches_with_network_id(address: bytes, network_id: int) -> bool:
+    return (address[0] & 0x0F) == network_id
 
-    # todo: GK - verify this is indeed correct (after IOHK confirmation/testnet support)
-    # protocol magic is included in Byron addresses only on testnets
-    if protocol_magic == ProtocolMagics.MAINNET:
-        address_attributes = {}
-        address_payload = None
+
+def has_byron_address_header(address: bytes) -> bool:
+    return (address[0] >> 4) == AddressId.BYRON_ADDRESS
+
+
+def get_human_readable_address(address: bytes, is_byron_address: bool) -> str:
+    if is_byron_address:
```

Comment:
Does it make sense to show byron addresses as bech32? If, not, I would use the address header to determine the address encoding so the function caller does not need to bother with that

Reply:
Good point. Fixed.